### PR TITLE
docs(review): rewrite stack-aware review spec + plan to deferral markers (#224)

### PR DESCRIPTION
**Pull request by howarewoo**
## Goal

Replace the descendant-PR-fetch design for stack-aware review (the now-closed #276–#278) with an inline **deferral-marker** approach: review honors a `woostack-defer(<ref>)` marker read from the PR's *own* diff, so the cost of declaring intentionally-deferred work moves upstream to plan/execute and every review stops fetching sibling PRs.

## Summary

- Rewrote the **spec** to the marker design — AC1–6, hardened Q1–8: co-located + reason match (anti-mis-fire, chunk-safe), `security`/wrong-in-PR-code/bare-`TODO` guards, `deferred_to` field survival verified against the `dict(df)` merge, execute self-clean + `woostack-status` surfacing for stale markers, `review.defer_markers` off-switch.
- Rewrote the **plan** to 3 increments / 8 tasks: Inc 1 review honors the marker (`defer_markers` key, classifier demotion + `deferred_count`, defender directive, render/docs) → Inc 2 plan declares / execute resolves + self-cleans / status surfaces → Inc 3 review `SKILL.md`.
- **Supersedes** the closed diff-verify stack: no `detect-stack.sh`, no `stack.md`, no `gh pr list`/`gh pr diff` of descendants — the signal lives in the diff.

### Automated

- [ ] Not run — docs-only change (spec + plan markdown). The plan's code tasks (load-config, intersect) ship their own tests in the follow-up increment PRs, not here.

### Manual

**Before merge**

- [ ] Read spec §4.0–§4.2 + §9 (Q1–8); confirm the marker design + lifecycle (declare → honor → resolve → surface) are sound.
- [ ] Confirm the plan's 8 tasks map to the spec ACs (plan Self-review) and the `woostack-defer` token parity holds across the 5 skills.

Spec: .woostack/specs/2026-06-09-review-stack-aware.md

### CODE REVIEW

**Comment by howarewoo** [comment]
* **#283** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/283?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/howarewoo/woostack/283?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: skills/woostack-review/prompts/_header.md [hunk 1]
```diff
@@ -243,6 +246,9 @@ for f in findings:
     blocking = bool(f.get("blocking", False))
 
     body = f"**{title}**\n\n{description}"
+    dt = (f.get("deferred_to") or "").strip()
+    if dt:
+        body += f"\n\n_Deferred to {dt} — a later increment completes this; non-blocking._"
```
Comment (howarewoo) [inline skills/woostack-review/prompts/_header.md:256]: **Nit: Unsanitized deferred_to injected into comment Markdown**

The body builder renders `deferred_to` verbatim into a GitHub PR comment as `_Deferred to {dt} — ...`. The validator directive (validator.md step 4b) explicitly tells the LLM to set `deferred_to` to the marker `<ref>` verbatim from the PR diff — content authored by the PR submitter. A crafted `<ref>` such as `x_](https://phishing.example.com` or `x\`code\`` can break out of the surrounding `_..._` italic delimiters and inject Markdown constructs (links, code spans, further italic-close pairs) into the rendered review comment. All other body-builder fields (`title`, `description`, `fix`, `suggestion`) are either LLM-generated (less directly attacker-controlled) or have explicit sanitization guards (the backtick fence guard on `suggestion`). The `deferred_to` path has no equivalent guard despite being the only field sourced verbatim from the untrusted diff.

Fix: Sanitize `dt` before embedding it. At minimum, escape Markdown control characters that are meaningful in the surrounding italic context: strip or replace `_`, `[`, `]`, `` ` ``, `*`, and `<` with their HTML-entity or escaped equivalents before interpolating. A simple approach consistent with the existing backtick guard pattern: `dt = re.sub(r'[_*\[\]`<>]', '', dt)` (strip to plain-text `<ref>`, since valid `<ref>` values like `increment 3` or `#225` contain none of these characters). Alternatively, avoid the inline italic context entirely by rendering the deferral note with HTML: `<em>Deferred to {html.escape(dt)} — a later increment completes this; non-blocking.</em>`.

<sub>— <strong>MEDIUM · NIT</strong> · flagged by the <code>security</code> agent</sub>

File: skills/woostack-plan/SKILL.md [hunk 2]
```diff
@@ -64,6 +64,27 @@ under the target and propose a further split; genuinely atomic changes may excee
...
+
+1. In the **deferring** increment, a step that drops a deferral marker at the gap site —
+   `woostack-defer(increment N): <reason>` — in the file's comment syntax (e.g.
+   `// woostack-defer(increment 3): call sites wired in increment 3`). The literal token is
+   `woostack-defer`; `<ref>` is the increment that completes the work.
+2. In the **implementing** increment (N), a step that **removes** that marker as part of wiring the
+   work, so the marker exists exactly while the gap is open.
+
+The marker is the single signal `woostack-review` reads to demote a "missing X" finding to a
+non-blocking `Deferred to N` nit (see [`woostack-review`](../woostack-review/SKILL.md) for the
```
Comment (howarewoo) [inline skills/woostack-plan/SKILL.md]: **Nit: Inconsistent rendered nit label: N vs <ref>**

The new Deferral markers section says the review gate demotes a finding to a non-blocking `Deferred to N` nit. But `woostack-review/SKILL.md` and `_header.md` consistently use `Deferred to <ref>` — where `<ref>` is the full marker reference (e.g. `"increment 3"`), not a bare number. `woostack-execute/SKILL.md` line 134 has the same mismatch. Using `N` implies a numeric PR reference and contradicts the prose in both the spec (§4.0: `<ref>` prefers `increment N`) and the body builder that renders the value verbatim.

Fix: Replace `Deferred to N` with `Deferred to <ref>` in `woostack-plan/SKILL.md` (line 84) to match the terminology used in `woostack-review/SKILL.md` and `_header.md`.

<sub>— <strong>LOW · NIT</strong> · flagged by the <code>skills</code> agent</sub>

File: skills/woostack-review/scripts/intersect-findings.sh [hunk 3]
```diff
@@ -597,13 +620,14 @@ cp "$FINAL" "$PRECLASSIFY"   # pre-floor snapshot for per-angle disagreement met
 classify_floor
 kept_count="$(jq 'length' "$FINAL")"
 nit_count="$(jq '[.[] | select(.nit == true)] | length' "$FINAL")"
+deferred_count="$(jq '[.[] | select((.deferred_to // "") != "" and .nit == true)] | length' "$FINAL")"
```
Comment (howarewoo) [inline skills/woostack-review/scripts/intersect-findings.sh:623]: **Nit: Adversarial-path deferred_count wiring never tested**

Every test case in test-intersect-deferred.sh sets disable_adversarial:true, so the adversarial code path (intersect-findings.sh lines 620-630) is never exercised. The deferred_count computation at line 623 and its write_metrics call at line 630 have zero test coverage. A bug in that code path — e.g., the metric argument position being off in the 10-arg write_metrics call — would pass the test suite undetected.

Fix: Add a fourth setup that provides both findings.defender.json and findings.prosecutor.json (with the same finding so it survives intersection), and omits disable_adversarial (or sets it false), then asserts deferred_count=1 in validator-metrics.json after running intersect-findings.sh.

<sub>— <strong>MEDIUM · NIT</strong> · flagged by the <code>tests</code> agent</sub>

File: skills/woostack-review/scripts/tests/test-intersect-deferred.sh [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+# severity_floor. Defender-only (disable_adversarial) isolates the floor classifier.
+setup() { # $1 = severity_floor ; $2 = defer_markers ("true"/"false")
+  work="$(mktemp -d)"
+  export OUTDIR="$work"
+  printf '{"disable_adversarial":true,"severity_floor":"%s","defer_markers":%s}\n' "$1" "$2" > "$OUTDIR/config.json"
+  cat > "$OUTDIR/findings.defender.json" <<'JSON'
+[
+  {"angle":"bugs","file":"x.ts","line":3,"severity":"HIGH","blocking":true,
+   "title":"Missing call-site wiring","description":"d","fix":"f","fix_type":"prose",
+   "suggestion":null,"rule_quote":null,"deferred_to":"increment 3"}
```
Comment (howarewoo) [inline skills/woostack-review/scripts/tests/test-intersect-deferred.sh:19]: **Nit: Empty/whitespace deferred_to edge case untested**

All three test fixtures set deferred_to to a non-empty string. The production guard at intersect-findings.sh:326 is `defer_on and isinstance(dt, str) and dt.strip()` — an empty string or whitespace-only deferred_to must NOT trigger demotion. This edge case is unexercised, so a regression that removed the .strip() guard would not be caught.

Fix: Add a case where the finding has deferred_to set to an empty string or whitespace (e.g. "deferred_to":"") and assert that nit stays false and blocking stays true — confirming the guard works.

<sub>— <strong>LOW · NIT</strong> · flagged by the <code>tests</code> agent</sub>

File: skills/woostack-review/scripts/tests/test-load-config-defer-markers.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+setup() { # $1 = config json body
+  work="$(mktemp -d)"
+  export OUTDIR="$work/out"
+  export GITHUB_WORKSPACE="$work/repo"
+  mkdir -p "$OUTDIR" "$GITHUB_WORKSPACE/.woostack"
+  printf '%s\n' "$1" > "$GITHUB_WORKSPACE/.woostack/config.json"
+}
+
+# defer_markers:false accepted + emitted to canonical config.
```
Comment (howarewoo) [inline skills/woostack-review/scripts/tests/test-load-config-defer-markers.sh:17]: **Nit: No test for defer_markers absent from config**

The three test cases cover false, true, and a non-boolean value, but not the omitted case. When defer_markers is absent from config.json, load-config.sh should not write the key to the output config (the consumer intersect-findings.sh re-defaults to true on its own). Verifying the absent-key behavior is cheap and guards against a regression where load-config.sh writes a spurious default value that could shadow the consumer's own defaulting logic.

Fix: Add a case using setup '{"review":{}}' and assert that jq -r '.defer_markers' "$OUTDIR/config.json" returns null (key absent), confirming load-config.sh does not inject a default value.

<sub>— <strong>LOW · NIT</strong> · flagged by the <code>tests</code> agent</sub>

File: .woostack/plans/2026-06-09-review-stack-aware.md [hunk 6]
```diff
@@ -40,449 +40,89 @@ setup() { # $1 = config json body
...
+#                                   default true. false ignores the markers.)
 ```
 
-- [ ] **Step 4: Run the test, confirm it passes**
+- [x] **Step 4: Run the test, confirm it passes**
 
-Run: `bash skills/woostack-review/scripts/tests/test-load-config-stack-aware.sh`
-Expected: PASS — `3 passed, 0 failed`
+Run: `bash skills/woostack-review/scripts/tests/test-load-config-defer-markers.sh`
+Expected: PASS — `3 passed, 0 failed`. Regression: the existing load-config suite still passes.
```
Comment (howarewoo) [inline .woostack/plans/2026-06-09-review-stack-aware.md]: **Nit: Plan test count wrong: claims 3 passed but test has 4 assertions**

The plan's Task 1 Step 4 claims `test-load-config-defer-markers.sh` produces `3 passed, 0 failed`, but the shipped test file contains 4 assertions: assert_eq for defer_markers:false, assert_eq for defer_markers:true, assert_eq for nonzero exit code, and assert_contains for the error message naming the key. The plan's own Step 3 code snippet also shows 4 assertions. A developer running the test after following the plan would see `4 passed, 0 failed` rather than the claimed `3 passed, 0 failed`.

Fix: Update line 108 to read: `Expected: PASS — \`4 passed, 0 failed\`. Regression: the existing load-config suite still passes.`

```suggestion
Expected: PASS — `4 passed, 0 failed`. Regression: the existing load-config suite still passes.
```

<sub>— <strong>LOW · NIT</strong> · flagged by the <code>docs</code> agent</sub>

File: .woostack/plans/2026-06-09-review-stack-aware.md [hunk 7]
```diff
@@ -491,60 +131,93 @@ ROOT="$(cd "$DIR/../../.." && pwd)"
...
 
-# A finding carrying stack_deferred must become a non-blocking nit regardless of
-# severity_floor. Run defender-only (disable_adversarial) to isolate the floor
-# classifier from the intersection logic.
-setup() { # $1 = severity_floor
+# A finding carrying deferred_to must become a non-blocking nit regardless of
+# severity_floor. Defender-only (disable_adversarial) isolates the floor classifier.
+setup() { # $1 = severity_floor ; $2 = defer_markers ("true"/"false")
   work="$(mktemp -d)"
   export OUTDIR="$work/out"
```
Comment (howarewoo) [inline .woostack/plans/2026-06-09-review-stack-aware.md]: **Nit: Plan snippet uses OUTDIR=$work/out but shipped test uses $work**

The plan's Task 2 code snippet shows `export OUTDIR="$work/out"` with a separate `mkdir -p "$OUTDIR"`, but the shipped test file `test-intersect-deferred.sh` uses `export OUTDIR="$work"` (no `/out` subdirectory). This matches the established pattern in other intersect tests such as `test-intersect-nits.sh`. A developer using the plan snippet as a reference for writing future tests would produce code inconsistent with the shipped file and the existing test suite.

Fix: Update the plan's test snippet at line 138 to use `export OUTDIR="$work"` (removing `/out`), matching the shipped test file and the established pattern in other intersect tests.

<sub>— <strong>LOW · NIT</strong> · flagged by the <code>docs</code> agent</sub>

**Review by howarewoo — COMMENTED** [review]
## AI Deep Code Review Summary

Reviewed the stack-aware deferral-marker rewrite — the spec/plan plus the review pipeline's intersect/load-config scripts, tests, and SKILL/_header/validator docs. No blocking issues. Seven non-blocking nits: a Markdown-injection hardening note on the verbatim `deferred_to` render path, three test-coverage gaps (the adversarial `deferred_count` path is never exercised, empty/whitespace `deferred_to` and absent-`defer_markers` cases untested), and doc/label consistency nits in the plan and `Deferred to <ref>` wording.

---
**Status: APPROVED** — No blocking findings, 7 nit(s). See inline comments.
*Audited by woostack-review · Host: Claude Code · Provider: anthropic · Model: opus/sonnet/haiku (per-angle tiers)*



_Review event downgraded to COMMENT — GitHub blocks APPROVE on your own PR. Status line above carries the actual verdict._

File: skills/woostack-review/prompts/_header.md (hunk 1 above)
Comment (howarewoo) [reply skills/woostack-review/prompts/_header.md:256]: Fixed in 008a32c: the body builder now strips Markdown control chars (` _ * [ ] < >) from `deferred_to` via `re.sub(r"[`_*\[\]<>]", "", dt)` before interpolation. Valid refs (`increment 3`, `#225`) pass through unchanged; breakout payloads like `x_](http://evil` are neutralized.

File: skills/woostack-plan/SKILL.md (hunk 2 above)
Comment (howarewoo) [reply skills/woostack-plan/SKILL.md]: Fixed in 008a32c: `Deferred to N` → `Deferred to <ref>` in woostack-plan/SKILL.md and woostack-execute/SKILL.md, matching the canonical wording in woostack-review/SKILL.md and _header.md.

File: skills/woostack-review/scripts/intersect-findings.sh (hunk 3 above)
Comment (howarewoo) [reply skills/woostack-review/scripts/intersect-findings.sh:623]: Fixed in 008a32c: added an adversarial case to test-intersect-deferred.sh — both validator passes present, `disable_adversarial` unset — asserting `mode=adversarial`, demotion to nit, and `deferred_count=1`. Exercises the intersect-findings.sh:623/630 path. Suite now 14 passed.

File: skills/woostack-review/scripts/tests/test-intersect-deferred.sh (hunk 4 above)
Comment (howarewoo) [reply skills/woostack-review/scripts/tests/test-intersect-deferred.sh:19]: Fixed in 008a32c: added a blank-`deferred_to` case ("   ") asserting the HIGH finding stays blocking and `deferred_count=0` — covers the `dt.strip()` guard at intersect-findings.sh:326.

File: skills/woostack-review/scripts/tests/test-load-config-defer-markers.sh (hunk 5 above)
Comment (howarewoo) [reply skills/woostack-review/scripts/tests/test-load-config-defer-markers.sh:17]: Fixed in 008a32c: added a `{"review":{}}` case asserting `.defer_markers == null` (loader writes no key when absent, so it can't shadow the consumer's default). Suite now 5 passed.

File: .woostack/plans/2026-06-09-review-stack-aware.md (hunk 6 above)
Comment (howarewoo) [reply .woostack/plans/2026-06-09-review-stack-aware.md]: Fixed in 008a32c: corrected to `5 passed` — the absent-key case from the sibling thread adds a 5th assertion. plan line 269 likewise bumped to `14 passed` for the intersect suite, since the new deferral cases land in the same PR.

File: .woostack/plans/2026-06-09-review-stack-aware.md (hunk 7 above)
Comment (howarewoo) [reply .woostack/plans/2026-06-09-review-stack-aware.md]: Fixed in 008a32c: the Task-2 snippet now uses `export OUTDIR="$work"` (mkdir dropped), matching the shipped test-intersect-deferred.sh and the `$work` convention in the other intersect tests.

File: skills/woostack-review/prompts/_header.md [hunk 8]
```diff
@@ -243,6 +246,14 @@ for f in findings:
     blocking = bool(f.get("blocking", False))
 
     body = f"**{title}**\n\n{description}"
+    dt = (f.get("deferred_to") or "").strip()
+    # Defense-in-depth: deferred_to is the only body field taken verbatim from the
+    # untrusted diff (the woostack-defer marker <ref>). Strip Markdown control
+    # chars so a crafted <ref> cannot break out of the italic note into links/code
+    # spans. Valid refs ("increment 3", "#225") are unaffected.
+    dt = re.sub(r"[`_*\[\]<>]", "", dt)
```
Comment (howarewoo) [inline skills/woostack-review/prompts/_header.md]: **Nit: Sanitizer omits newlines — italic span can break**

The new `re.sub(r"[`_*\[\]<>]", "", dt)` strips Markdown control characters from `deferred_to` but does not strip newlines (`\n`). The value is embedded in a single-line italic span: `_Deferred to {dt} — ..._`. A `deferred_to` string containing a newline (e.g. `"increment 3\nmalicious text"`) breaks the italic span mid-line, causing the remainder to render as unstyled text outside the emphasis — defeating the stated defense-in-depth goal.

Fix: Add `\n` (and optionally `\r`) to the stripped character class: `re.sub(r"[`_*\[\]<>\n\r]", "", dt)`. This is consistent with the comment's intent (prevent break-out of the italic note) and does not affect valid refs like `"increment 3"` or `"#225"`.

```suggestion
    dt = re.sub(r"[`_*\[\]<>\n\r]", "", dt)
```

<sub>— <strong>MEDIUM · NIT</strong> · flagged by the <code>bugs</code> agent</sub>

**Review by howarewoo — COMMENTED** [review]
## AI Deep Code Review Summary

Incremental recheck of commit `008a32c` (the fix addressing the prior 7 nits). All prior threads are resolved and the security/test/doc fixes hold — the only new finding is one follow-on nit: the `deferred_to` sanitizer strips Markdown control chars but not interior newlines, leaving a residual block-injection gap (`\n` + `#`/`-`).

---
**Status: APPROVED** — No blocking findings, 1 nit(s). See inline comments.
*Audited by woostack-review · Host: Claude Code · Provider: anthropic · Model: opus/sonnet/haiku (per-angle tiers) · incremental*



_Review event downgraded to COMMENT — GitHub blocks APPROVE on your own PR. Status line above carries the actual verdict._

File: skills/woostack-review/prompts/_header.md (hunk 8 above)
Comment (howarewoo) [reply skills/woostack-review/prompts/_header.md]: Fixed in 496bfdb: added `\n\r` to the sanitizer char class — `re.sub(r"[`_*\[\]<>\n\r]", "", dt)`. An interior newline is now stripped, so `increment 3\n# heading` collapses to one line and can't inject a block element. Valid refs (`increment 3`, `#225`) unaffected.
